lib/src/observable.dart

```diff
@@ -1,92 +1,131 @@
-    final callback = Pointer.fromFunction<obx_observer>(_anyCallback);
-    final storePtr = store.ptr;
-    _anyObserver[storePtr.address] =
-        bindings.obx_observe(storePtr, callback, storePtr.cast<Void>());
-    StoreCloseObserver.addListener(store, _anyObserver[storePtr.address], () {
-      unsubscribe(store);
-    });
+  _Observer() {
+    initializeDartAPI();
+  }
+
+  // start() is called whenever user starts listen()-ing to the stream
+  void init(void Function() start) {
+    controller = StreamController<StreamValueType>(
+        onListen: start, onPause: stop, onResume: start, onCancel: stop);
   }
 
-  // #53 ffi:Pointer finalizer
-  static void unsubscribe(Store store) {
-    final storeAddress = store.ptr.address;
-    if (!_anyObserver.containsKey(storeAddress)) {
-      return;
-    }
-    StoreCloseObserver.removeListener(store, _anyObserver[storeAddress]);
-    bindings.obx_observer_close(_anyObserver[storeAddress]);
-    _anyObserver.remove(storeAddress);
-    syncOrObserversExclusive.unmark(store);
+  // stop() is called when the stream subscription is paused or canceled
+  void stop() {
+    _debugLog('stopped');
+    if (_cObserver != null) checkObx(bindings.obx_observer_close(_cObserver));
   }
 
-  static bool isSubscribed(Store store) =>
-      _Observable._anyObserver.containsKey(store.ptr.address);
+  void _debugLog(String message) {
+    // print('Observer=${_cObserver?.address} ' + message);
+  }
 }
 
-extension Streamable<T> on Query<T> {
-  void _setup() {
-    if (!_Observable.isSubscribed(store)) {
-      _Observable.subscribe(store);
-    }
-    final storeAddress = store.ptr.address;
-
-    _Observable._any[storeAddress] ??= <int, Any>{};
-    _Observable._any[storeAddress] /*!*/ [entityId] ??= (u, _, __) {
-      // dummy value to trigger an event
-      _Observable.controller.add(entityId);
-    };
+/// StreamController implementation inspired by the sample controller sample at:
+/// https://dart.dev/articles/libraries/creating-streams#honoring-the-pause-state
+/// https://dart.dev/articles/libraries/code/stream_controller.dart
+extension ObservableStore on Store {
+  /// Create a stream to data changes on EntityT (stored Entity class).
+  ///
+  /// The stream receives an event whenever an object of EntityT is created or
+  /// changed or deleted. Make sure to close() the subscription after you're
+  /// done with it to avoid hanging change listeners.
+  Stream<void> subscribe<EntityT>() {
+    final observer = _Observer<void>();
+    final entityId = entityDef<EntityT>().model.id.id;
+
+    observer.init(() {
+      // We're listening to events on single entity so there's no argument.
+      // Ideally, controller.add() would work but it doesn't, even though we're
+      // using StreamController<Void> so the argument type is `void`.
+      observer.receivePort = ReceivePort()
+        ..listen((_) => observer.controller.add(null));
+      observer.cObserver = bindings.obx_dart_observe_single_type(
```

Comment:
I think you should close previous if `cObserver != null`. this is if you want to subscribe multiple times to the same thread.In this case, `onListen` will be called several times and` onCancel` only when the last one is unsubscribed.

Reply:
same as above: https://github.com/objectbox/objectbox-dart/pull/145#discussion_r535180104